### PR TITLE
Add `created_at` and `updated_at` fields

### PR DESCRIPTION
**Pull request by fivetran-avinash**
## PR Overview
**This PR will address the following Issue/Feature:** [#144]

**This PR will result in the following new package version:** v0.17.0

Although this should not impact the existing fields, the general ledger model now is different, so I decided to bump it up. I'm open to moving it back down to 0.16.1 if we don't feel the change is significant enough. 

**Please provide the finalized CHANGELOG entry which details the relevant changes included in this PR:**

## Feature Updates
- We have introduced the following fields in the `quickbooks__general_ledger`.
  - `created_at`: The time a transaction was first created.
  - `updated_at`: The time a transaction was last updated.
  - This required updates to all our double entry transaction models to bring in these specific fields for each transaction type. 
- This isn't necessarily a breaking change, but as this is a new version of this model that introduces new fields, we determined it was best to bump the version up.

### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt run –full-refresh && dbt test
- [NA] dbt run (if incremental models are present) && dbt test

Before marking this PR as "ready for review" the following have been applied:
- [x] The appropriate issue has been linked, tagged, and properly assigned
- [x] All necessary documentation and version upgrades have been applied
- [x] docs were regenerated (unless this PR does not include any code or yml updates)
- [x] BuildKite integration tests are passing
- [x] Detailed validation steps have been provided below

### Detailed Validation
Please share any and all of your validation steps:

Confirmed that `created_at` and `updated_at` are brought into the general ledger model. 

Validation tests passed with integration tests seed data.

Screenshot 2024-11-20 at 7 35 55 AM

### If you had to summarize this PR in an emoji, which would it be?
 
⌛

### CODE REVIEW

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-avinash this is looking great! I just have a few small requests and one small question before approving.

File: integration_tests/dbt_project.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 name: 'quickbooks_integration_tests'
 
-version: '0.16.0'
+version: '0.17.0'
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply integration_tests/dbt_project.yml:3]: This must be an artifact from testing on seed data. Removing from the actual model.

File: README.md [hunk 2]
```diff
@@ -73,7 +73,7 @@ Include the following QuickBooks package version in your `packages.yml` file.
 ```yaml
 packages:
   - package: fivetran/quickbooks
-    version: [">=0.16.0", "<0.17.0"] # we recommend using ranges to capture non-breaking changes automatically
+    version: [">=0.17.0", "<0.18.0"] # we recommend using ranges to capture non-breaking changes automatically
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply README.md:76]: Updated

File: packages.yml [hunk 3]
```diff
@@ -1,3 +1,4 @@
 packages:
-  - package: fivetran/quickbooks_source
-    version: [">=0.11.0", "<0.12.0"]
\ No newline at end of file
+  - git: https://github.com/fivetran/dbt_quickbooks_source.git
+    revision: feature/add-created-at-and-updated-at
+    warn-unpinned: false
```
Comment (fivetran-avinash) [inline packages.yml:4]:
```suggestion
  - package: fivetran/quickbooks_source
    version: [">=0.12.0", "<0.13.0"]
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,13 @@
+# dbt_quickbooks v0.17.0
+[PR #146](https://github.com/fivetran/dbt_quickbooks/pull/146) introduces the following updates:
+
+## Feature Updates
+- We have introduced the following fields in the `quickbooks__general_ledger` to better analyze real-time transaction data:
+  - `created_at`: The time a transaction was first created.
+  - `updated_at`: The time a transaction was last updated.
+  - This required updates to all our double entry transaction models to bring in these specific fields for each transaction type. 
+- This isn't necessarily a breaking change, but as this introduces new fields to an end model, we determined it was best to bump the version up.
```
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply CHANGELOG.md]: Removed

File: packages.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (fivetran-avinash) [reply packages.yml:4]: Added a suggestion reminder

**Review by fivetran-avinash — COMMENTED** [review]
@fivetran-joemarkiewicz This is ready for re-review.

**Review by fivetran-joemarkiewicz — APPROVED** [review]
LGTM

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,8 +1,12 @@
-# dbt_quickbooks v0.17.0-a1
-
-## Bug Fix
-- Updated the logic in `int_quickbooks__retained_earnings` to ensure accounting periods with no revenue and expense class lines were accounted for. 
-  - This will ensure the net income adjustment is available regardless of existing revenue or expenses.
+# dbt_quickbooks v0.17.0
+[PR #146](https://github.com/fivetran/dbt_quickbooks/pull/146) introduces the following updates:
+
+## Breaking Changes
+- We have introduced the following fields in the `quickbooks__general_ledger` to better analyze real-time transaction data:
```
Comment (fivetran-catfritz) [inline CHANGELOG.md:5]:
```suggestion
- Introduced the following fields in the `quickbooks__general_ledger` model to better analyze real-time transaction data:
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,8 +1,12 @@
...
+  - `created_at`: The time a transaction was first created.
+  - `updated_at`: The time a transaction was last updated.
+  - This required updates to all our double entry transaction models to bring in these specific fields for each transaction type.
+- This is a breaking change as this adds new fields to the existing schema for one of our end models.
```
Comment (fivetran-catfritz) [inline CHANGELOG.md:9]:
```suggestion
- This is a breaking change as this adds new fields to the existing schema. 
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,8 +1,12 @@
...
-- Updated the logic in `int_quickbooks__retained_earnings` to ensure accounting periods with no revenue and expense class lines were accounted for. 
-  - This will ensure the net income adjustment is available regardless of existing revenue or expenses.
+# dbt_quickbooks v0.17.0
+[PR #146](https://github.com/fivetran/dbt_quickbooks/pull/146) introduces the following updates:
+
+## Breaking Changes
+- We have introduced the following fields in the `quickbooks__general_ledger` to better analyze real-time transaction data:
+  - `created_at`: The time a transaction was first created.
+  - `updated_at`: The time a transaction was last updated.
+  - This required updates to all our double entry transaction models to bring in these specific fields for each transaction type.
```
Comment (fivetran-catfritz) [inline CHANGELOG.md:8]:
```suggestion
  - Updated the `*_double_entry` models to add these fields for each transaction type.
```

File: models/intermediate/int_quickbooks__retained_earnings.sql [hunk 8]
```diff
@@ -4,19 +4,42 @@ with general_ledger_balances as (
...
-        sum(case when account_class = 'Revenue' then period_net_change else 0 end) as revenue_net_change,
-        sum(case when account_class = 'Revenue' then period_net_converted_change else 0 end) as revenue_net_converted_change,
-        sum(case when account_class = 'Expense' then period_net_change else 0 end) as expense_net_change,
-        sum(case when account_class = 'Expense' then period_net_converted_change else 0 end) as expense_net_converted_change
+        sum(period_net_change) as revenue_net_change,
+        sum(period_net_converted_change) as revenue_net_converted_change
     from general_ledger_balances
+    
+    where account_class = 'Revenue'
+
```
Comment (fivetran-catfritz) [inline models/intermediate/int_quickbooks__retained_earnings.sql:17]: Can you explain these changes in the changelog? It's just not obviously related to the created/update_at if it's part of those changes.

**Review by fivetran-catfritz — COMMENTED** [review]
@fivetran-avinash a couple comments/questions!

File: CHANGELOG.md (hunk 5 above)
Comment (fivetran-avinash) [reply CHANGELOG.md]: Committed!

File: CHANGELOG.md (hunk 6 above)
Comment (fivetran-avinash) [reply CHANGELOG.md]: Changed.

File: CHANGELOG.md (hunk 7 above)
Comment (fivetran-avinash) [reply CHANGELOG.md]: Updated.

**Review by fivetran-avinash — COMMENTED** [review]
@fivetran-catfritz Addressed your comments!

File: models/intermediate/int_quickbooks__retained_earnings.sql (hunk 8 above)
Comment (fivetran-catfritz) [reply models/intermediate/int_quickbooks__retained_earnings.sql:17]: Got it. I'll leave it to your judgement!

**Review by fivetran-catfritz — APPROVED** [review]
@fivetran-avinash lgtm!

**Pull request merged**